matrix/src/lib.rs

```diff
@@ -92,66 +97,49 @@ pub trait MatrixRows<T>: Matrix<T> {
         )
     }
 
-    fn vertically_strided(self, stride: usize, offset: usize) -> VerticallyStridedMatrixView<Self>
+    fn horizontally_packed_row<'a, P>(
+        &'a self,
+        r: usize,
+    ) -> (impl Iterator<Item = P>, impl Iterator<Item = T>)
     where
-        Self: Sized,
+        P: PackedValue<Value = T>,
+        T: Clone + 'a,
     {
-        VerticallyStridedMatrixView {
-            inner: self,
-            stride,
-            offset,
-        }
+        let num_packed = self.width() / P::WIDTH;
+        let packed = (0..num_packed).map(move |c| P::from_fn(|i| self.get(r, P::WIDTH * c + i)));
+        let sfx = (num_packed * P::WIDTH..self.width()).map(move |c| self.get(r, c));
+        (packed, sfx)
     }
-}
-
-/// A `Matrix` which supports access its rows as slices.
-pub trait MatrixRowSlices<T>: MatrixRows<T> {
-    fn row_slice(&self, r: usize) -> &[T];
 
-    fn row_slices<'a>(&'a self) -> impl Iterator<Item = &'a [T]>
+    fn vertically_packed_row<P>(&self, r: usize) -> impl Iterator<Item = P>
     where
-        T: 'a,
+        P: PackedValue<Value = T>,
     {
-        (0..self.height()).map(|r| self.row_slice(r))
+        (0..self.width()).map(move |c| P::from_fn(|i| self.get(r + i, c)))
     }
-}
 
-/// A `Matrix` which supports access its rows as mutable slices.
-pub trait MatrixRowSlicesMut<T>: MatrixRowSlices<T> {
-    fn row_slice_mut(&mut self, r: usize) -> &mut [T];
-
-    // BEWARE: if we add a matrix type that has several rows in the same memory location,
-    // these default implementations will be invalid
-    // For example, a "tiling" matrix view that repeats its rows
-
-    /// # Safety
-    /// Each row index in `rs` must be unique.
-    unsafe fn disjoint_row_slices_mut<const N: usize>(&mut self, rs: [usize; N]) -> [&mut [T]; N] {
-        rs.map(|r| {
-            let s = self.row_slice_mut(r);
-            // launder the lifetime to 'a instead of being bound to self
-            unsafe { core::slice::from_raw_parts_mut(s.as_mut_ptr(), s.len()) }
-        })
-    }
-    fn row_pair_slices_mut(&mut self, r0: usize, r1: usize) -> (&mut [T], &mut [T]) {
-        // make it safe by ensuring rs unique
-        assert_ne!(r0, r1);
-        let [s0, s1] = unsafe { self.disjoint_row_slices_mut([r0, r1]) };
-        (s0, s1)
+    fn vertically_strided(self, stride: usize, offset: usize) -> VerticallyStridedMatrixView<Self>
+    where
+        Self: Sized,
+    {
+        VerticallyStridedPerm::new_view(self, stride, offset)
     }
-}
 
-pub trait MatrixRowChunksMut<T: Send>: MatrixRowSlicesMut<T> {
-    type RowChunkMut<'a>: MatrixRowSlicesMut<T> + Send
+    fn columnwise_dot_product<EF>(&self, v: &[EF]) -> Vec<EF>
```

Comment:
nit: let's add a comment, might not be totally clear from the name what it does. I think one way we could describe it is "premultiply this matrix by the given vector"